api/api.go

```diff
@@ -112,3 +125,342 @@ func (api *API) statsHandler(ctx router.HTTPContext) {
+		return
+	}
+
+	fmt.Fprintln(ctx.Response, string(out))
+}
+
+func putEntityHandler(ctx router.HTTPContext, c db.Connector) {
+	if len(ctx.Arguments) != 2 {
+		ctx.APIError(router.ErrMissingParam, "Missing param: id")
+		return
+	}
+
+	requestId := ctx.Arguments[1]
+	var requestUpdate db.Entity
+	err := json.NewDecoder(ctx.Request.Body).Decode(&requestUpdate)
+	if err != nil {
+		ctx.APIError(router.ErrInternal, "Can't decode request body (error: %s)", err)
+		return
+	}
+
+	q := queryForId(requestId)
+
+	entities, err := c.UpdateEntities(q, requestUpdate)
+	if err != nil {
+		if _, ok := err.(db.ErrUpdate); ok {
+			ctx.APIError(router.ErrInternal, "Error updating entity (error: %s)", err)
+		} else {
+			ctx.APIError(router.ErrInternal, "Uknown error when updating entity(error: %s)", err)
+		}
+
+		return
+	}
+
+	out, err := json.Marshal(entities[0])
+	if err != nil {
+		ctx.APIError(router.ErrInternal, "Can't encode response (error: %s)", err)
+		return
+	}
+
+	fmt.Fprintln(ctx.Response, string(out))
+}
+
+func deleteEntityHandler(ctx router.HTTPContext, c db.Connector) {
+	if len(ctx.Arguments) != 2 {
+		ctx.APIError(router.ErrMissingParam, "Missing param: id")
+		return
+	}
+
+	requestId := ctx.Arguments[1]
+	q := queryForId(requestId)
+
+	entities, err := c.DeleteEntities(q)
+	if err != nil {
+		if _, ok := err.(db.ErrDelete); ok {
+			ctx.APIError(router.ErrInternal, "Error deleting entity (error: %s)", err)
+		} else {
+			ctx.APIError(router.ErrInternal, "Uknown error when deleting entity(error: %s)", err)
+		}
+
+		return
+	}
+
+	out, err := json.Marshal(entities[0])
+	if err != nil {
+		ctx.APIError(router.ErrInternal, "Can't encode response (error: %s)", err)
+		return
+	}
+
+	fmt.Fprintln(ctx.Response, string(out))
+}
+
+func postEntitiesHandler(ctx router.HTTPContext, c db.Connector) {
+	var entities []db.Entity
+	err := json.NewDecoder(ctx.Request.Body).Decode(&entities)
+	if err != nil {
+		ctx.APIError(router.ErrInternal, "Can't decode request body (error: %s)", err)
+		return
+	}
+
+	for _, e := range entities {
+		ConvertFloat64ToInt(e)
+
+		for _, v := range e {
+			if !validValueType(v) {
+				ctx.APIError(router.ErrBadRequest, "Invalid Value Type: %v. Value type must be string or int", reflect.TypeOf(v))
+			}
+		}
+	}
+
+	ids, err := c.CreateEntities(entities)
+	if err != nil {
+		if _, ok := err.(db.ErrCreate); ok {
+			ctx.APIError(router.ErrInternal, "Error creating entity (error: %s)", err)
+		} else {
+			ctx.APIError(router.ErrInternal, "Uknown error when creating entity(error: %s)", err)
+		}
+
+		return
+	}
+
+	out, err := json.Marshal(ids)
+	if err != nil {
+		ctx.APIError(router.ErrInternal, "Can't encode response (error: %s)", err)
+		return
+	}
+
+	fmt.Fprintln(ctx.Response, string(out))
+}
+
+func getEntitiesHandler(ctx router.HTTPContext, c db.Connector) {
+	// Translate URL query to query struct
+	requestLabelSelector := ctx.Request.Form["query"][0]
```

Comment:
If `"query"` is missing, then `[0]` might cause a panic trying to access nil.

Reply:
ah good catch. updated the code + added test cases.